GH-80789: Get rid of the ``ensurepip`` infra for many wheels

webknjaz opened the pull request:
This is a refactoring change that aims to simplify ``ensurepip``. Before it, this module had legacy infrastructure that made an assumption that ``ensurepip`` would be provisioning more then just a single wheel. That assumption is no longer true since [[1]][[2]][[3]].

In this change, the improvement is done around removing unnecessary loops and supporting structures to change the assumptions to expect only the bundled or replacement ``pip`` wheel.

This is a spin-off of https://github.com/python/cpython/pull/12791.

[1]: https://github.com/python/cpython/commit/ece20db
[2]: https://github.com/python/cpython/pull/101039
[3]: https://github.com/python/cpython/issues/95299

<!-- gh-issue-number: gh-80789 -->
* Issue: gh-80789
<!-- /gh-issue-number -->

webknjaz committed:
Get rid of the ``ensurepip`` infra for many wheels

This is a refactoring change that aims to simplify ``ensurepip``.
Before it, this module had legacy infrastructure that made an
assumption that ``ensurepip`` would be provisioning more then just a
single wheel. That assumption is no longer true since [[1]][[2]][[3]].

In this change, the improvement is done around removing unnecessary
loops and supporting structures to change the assumptions to expect
only the bundled or replacement ``pip`` wheel.

[1]: https://github.com/python/cpython/commit/ece20db
[2]: https://github.com/python/cpython/pull/101039
[3]: https://github.com/python/cpython/issues/95299

vstinner commented on Lib/ensurepip/__init__.py:
```diff
@@ -27,8 +25,13 @@
 
-def _find_packages(path):
-    packages = {}
+def _find_packages(path: str | None) -> _Package:
```
If it returns a single package, it should be called _get_package(), singular, no?

webknjaz replied:
You're right. I was so focused on making the change as small as possible that I missed it. Thanks!

webknjaz replied:
Renamed to `_get_replacement_pip_package()`. Marking this thread as resolved.

vstinner commented on Misc/NEWS.d/next/Library/2023-09-10-23-35-20.gh-issue-109245.yCVQbo.rst:
```diff
@@ -0,0 +1,3 @@
+Simplified ``ensurepip`` to stop assuming that it can provision multiple
+wheels. The refreshed implementation now expects to only provision
+a ``pip`` wheel and no other distribution packages.
```
I don't think that a Changelog entry is needed. For end users, there is no visible change.

webknjaz replied:
Don't the downstream packagers deserve it?

webknjaz replied:
@vstinner would you apply the corresponding labels, then?

webknjaz committed:
Rename `_find_packages` to `_get_replacement_pip_package`

This change is intended to make it clear that the helper now only
returns one package and no more.

Co-Authored-By: vstinner@python.org

AA-Turner committed:
Remove changelog entry

AA-Turner committed:
Simplify WHEEL_PKG_DIR search

- Exit early if there are no files in the directory
- Return a match early by iterating in reverse order
- Merge filename checks
- Inline the path variable
- Remove type annotations

AA-Turner committed:
Simplify the pip wheel info

- Return a dict with consistent fields
- Remove caching
- Remove type annotations
- Leverage the known wheel package dir value to calculate full paths

vstinner reviewed (approved):
LGTM. But I would prefer to get a second review on this one.

pfmoore commented on Lib/ensurepip/__init__.py:
```diff
@@ -27,52 +17,36 @@
     # of the same package, but don't attempt to implement correct version
     # comparison since this case should not happen.
-    filenames = sorted(filenames)
+    filenames = sorted(filenames, reverse=True)
```
Why reverse? I guess it gives better odds of getting the correct version, but as the comment above says, we're not doing version comparison because it should never be needed.

webknjaz replied:
@AA-Turner ^

AA-Turner replied:
I noted this in the commit message but not a PR comment; sorry. The implementation of this function now short-circuits and returns the first match, but that means we need to reverse the order as previously the final match would be returned, and we want to preserve that behaviour. We retain no guarantees about finding the correct version.

pfmoore reviewed (approved):
One minor comment, but it doesn't block this - feel free to leave it if you think your version is better.

webknjaz commented on Lib/ensurepip/__init__.py:
```diff
@@ -27,52 +17,36 @@
     except OSError:
         # Ignore: path doesn't exist or permission error
-        filenames = ()
+        return None
```
@AA-Turner so the reason I was using a `LookupError` was that it's better structurally and using sentinels like `None` introduces a mental burden on the callers forcing them not to forget to check for a magic value of a different type with a special meaning. This reminds me of ancient languages or golang where this same burden looks like having to check some integer return value for whether it's `-1` to decide if it's usable.

AA-Turner replied:
No strong opinions here; I see both patterns. The caller must do something with either pattern -- either check for ``None`` or catch the error. No objections to go back to the error pattern if you'd prefer.

webknjaz replied:
I like forcing the caller to react rather than hoping that side won't forget... Anyway, I included this change as a part of the pathlib refactoring @ https://github.com/python/cpython/commit/b2a69a121a82d5abc8b99eed2492c4602d9fb91b. Marking the thread as resolved.

webknjaz commented on Lib/ensurepip/__init__.py:
```diff
@@ -27,52 +17,36 @@
 
-def _find_packages(path):
-    packages = {}
+def _find_wheel_pkg_dir_pip():
```
@AA-Turner I don't understand why you named it like this — it doesn't return a directory as far as I see. It looks like it returns a replacement package...

AA-Turner replied:
The function locates pip in the wheel package directory. I don't see it as "replacement" as the point of the ``--with-wheel-pkg-dir`` option is that distributors remove the bundled wheels in ``_bundled``, so there is no 'original' wheel to replace from that distributor's perspective.

webknjaz replied:
Well, until the downloader PR is merged, there's the original "source" wheel right in the Git repo, and when distros want to replace it, they add an additional source wheel, making it two (unless they also delete the first one, of course). Hence, replacement...

webknjaz replied:
I don't have better suggestions so I'll just mark this as resolved.

webknjaz commented on Lib/ensurepip/__init__.py:
```diff
@@ -27,52 +17,36 @@
     # comparison since this case should not happen.
-    filenames = sorted(filenames)
+    filenames = sorted(filenames, reverse=True)
     for filename in filenames:
```
@AA-Turner if we do want to include this much refactoring in the PR, why no just use my pathlib approach that reads better without looping and going into the lower-level abstractions, being shorter and easier to grasp:
```python
dist_matching_wheels = Path(_WHEEL_PKG_DIR).glob('pip-*.whl')
try:
    last_matching_dist_wheel = sorted(dist_matching_wheels)[-1]
except IndexError as index_err:
    raise LookupError from index_err
filename = last_matching_dist_wheel.name
version = filename.removeprefix('pip-').partition('-')[0]
```
_(https://github.com/python/cpython/pull/12791/files#diff-f5e321563d5bc98371a960662f654a9387fd24f9bacb382d5b90d417ad2569c7R197-R205)_

AA-Turner replied:
I like the ``glob`` approach (we could use ``glob.iglob`` or ``pathlib``).

webknjaz replied:
This is implemented in https://github.com/python/cpython/commit/b2a69a121a82d5abc8b99eed2492c4602d9fb91b, so I'm marking the thread as resolved.

AA-Turner commented on Lib/ensurepip/__init__.py:
```diff
@@ -167,24 +141,20 @@ def _bootstrap(*, root=None, upgrade=False, user=False,
+
+        filename = os.path.join(tmpdir, wheel_name)
+        with open(filename, "wb") as fp:
+            fp.write(whl)
```
If you change to using ``pathlib``, we could leverage the traversable API a little here:

```suggestion
        if package["bundled"]:
            # Use bundled wheel package
            wheel_path = resources.files("ensurepip") / "_bundled" / wheel_name
        else:
            # Use the wheel package directory
            wheel_path  = Path(_WHEEL_PKG_DIR, wheel_name)

        whl = wheel_path.read_bytes()
        filename = Path(tmpdir, wheel_name)
        filename .write_bytes(whl)
```

webknjaz replied:
I think that's what I've been doing in the original PR, just with a `nullcontext` for clever code linearization...
https://github.com/python/cpython/commit/c31bdca823eab7e4ffcd70c7bd9a9895506e1e33#diff-f5e321563d5bc98371a960662f654a9387fd24f9bacb382d5b90d417ad2569c7R238

webknjaz replied:
https://github.com/python/cpython/commit/b2a69a121a82d5abc8b99eed2492c4602d9fb91b implemented this in a more linear way. I'm marking the thread as resolved.

AA-Turner commented on Lib/ensurepip/__init__.py:
```diff
@@ -197,7 +167,8 @@ def _bootstrap(*, root=None, upgrade=False, user=False,
             args += ["-" + "v" * verbosity]
 
-        return _run_pip([*args, *_PACKAGE_NAMES], additional_paths)
+        return _run_pip([*args, "pip"], [filename])
```
(Following on from the above if ``pathlib`` is adopted)
```suggestion
        return _run_pip([*args, "pip"], [str(filename)])
```

pradyunsg replied:
Should that not be `os.fsdecode(filename)` instead?

AA-Turner replied:
Perhaps? I considered using ``os.fspath``, but ``Path`` just calls ``str`` for ``__fspath__`` internally so I took the shortcut. I'm not sure on ``fsdecode`` vs ``fspath`` though.

webknjaz replied:
Why decode and not encode, then? Don't we pass this over the OS API boundary into another process?

webknjaz replied:
Implemented as https://github.com/python/cpython/commit/b2a69a121a82d5abc8b99eed2492c4602d9fb91b. Marking as resolved.

webknjaz committed:
Use `pathlib` in `ensurepip` internally

It provides us with the ability to write simpler high-level logic that
is easier to understand. As a side effect, it became possible to make
the code less branchy.

webknjaz committed:
Drop `os.fsencode` from wheel path preprocessing

It was returning bytes on FreeBSD which was incorrectly injected into
the Python code snippet executed in a subprocess and had a b-preffix.

webknjaz committed:
Update replaced `_get_pip_whl_path_ctx` references

Some code comments and test function names were still referring to the
removed function name. Not anymore!

webknjaz committed:
Drop looping from the wheel verification script

This script is separate and is only used in CI as opposed to runtime.

pradyunsg commented on Lib/ensurepip/__init__.py:
```diff
@@ -171,7 +171,7 @@ def _bootstrap(*, root=None, upgrade=False, user=False,
             args += ["-" + "v" * verbosity]
 
-        return _run_pip([*args, "pip"], [os.fsencode(tmp_wheel_path)])
+        return _run_pip([*args, "pip"], [str(tmp_wheel_path)])
```
Use `fsdecode` for this?

webknjaz replied:
@pradyunsg
Could you explain why? It's made out of a pathlib object that already returns a normal unicode string. This is then injected into a snippet of Python code executed in a subprocess.

Is there a case when it'd matter? We don't work with the OS APIs on this level so I figure it'd do nothing..

pfmoore replied:
IMO, simply because it saves people having to reason about the code. `tmp_wheel_path` is a path object, and `os.fsdecode` converts path objects to strings safely, so we know this is safe, *without* having to think too hard about edge cases like undecodable bytes, surrogate escape characters, etc.

In this case yes, both will do the same thing[^1], but why make the reader work harder than necessary?

[^1]: I assume - I haven't thought it through, which is sort of my point 🙂

webknjaz replied:
I'm questioning this exactly because I'm not sure it'd do the right or equivalent thing. I don't mind integrating the suggestion, just want to understand it first.

pfmoore replied:
`os.fsdecode` will convert a bytes value correctly, whereas `str()` won't. It's not going to matter here, but `fsdecode` is, in some sense, "more correct".

In reality, though, the way to think of it is "`fsdecode` makes a filename (in whatever form you get it) into a string". And `fsencode` is the same, but to return bytes.

Honestly, though, I'm not that fussed. For this situation, `str` seems good enough even if `fsdecode` is technically better.

pradyunsg replied:
(I've had a travel day today, and am only just seeing this)

Thanks @pfmoore for the replies here -- I'm basically on the same page. `fsdecode` more clearly communicates what the intended behaviour is here and it's _technically_ better.

> I'm not sure it'd do the right or equivalent thing

It will be equivalent for str-based `Path`s. That function takes a path / path-like object and converts it to a string representation matching the filesystem encoding (which is relevant in the case of bytes paths, but we don't have that here).

It'll return a `str` unchanged.

webknjaz replied:
Makes sense. Will do.

pradyunsg commented on Lib/ensurepip/__init__.py:
```diff
@@ -14,39 +16,51 @@
-_WHEEL_PKG_DIR = sysconfig.get_config_var('WHEEL_PKG_DIR')
+_WHEEL_PKG_DIR = (
+    whl_pkg_dir_str := sysconfig.get_config_var('WHEEL_PKG_DIR')
+) is not None and Path(whl_pkg_dir_str).resolve() or None
```
This is fairly difficult to parse and understand correctly. There's no real strong reason to make this a single expression.

Here's a suggestion I have:

```suggestion
if (_pkg_dir := sysconfig.get_config_var('WHEEL_PKG_DIR')) is not None:
    _WHEEL_PKG_DIR = Path(_pkg_dir).resolve()
else:
    _WHEEL_PKG_DIR = None
```

pradyunsg commented on Lib/ensurepip/__init__.py:
```diff
@@ -14,39 +16,51 @@
 def _find_wheel_pkg_dir_pip():
     if _WHEEL_PKG_DIR is None:
-        return None
+        raise LookupError(
```
I'd much rather we continue returning `None` here rather than add complexity here with the suppress.

webknjaz replied:
I have a different preference, but will change.

webknjaz committed:
Multiline complex wheel path conditional

Co-authored-by: Pradyun Gedam <pradyunsg@gmail.com>

webknjaz committed:
Drag the wheel path object through `os.fsdecode`

webknjaz committed:
Use `None`-sentinel for missing alternative wheel

Co-authored-by: Pradyun Gedam <pradyunsg@gmail.com>

webknjaz committed:
Un-f-string alternative wheel path glob

It was tripping the linters and became unnecessary after hardcoding
the pip wheel filename prefix in the string.

webknjaz committed:
Melt the second wheel lookup error into sentinel

webknjaz commented on Lib/ensurepip/__init__.py:
```diff
@@ -1,78 +1,64 @@
+if (_pkg_dir := sysconfig.get_config_var('WHEEL_PKG_DIR')) is not None:
+    _WHEEL_PKG_DIR = Path(_pkg_dir).resolve()
+else:
+    _WHEEL_PKG_DIR = None
```
@pradyunsg converting this introduced a new uncovered line here. In https://github.com/python/cpython/pull/12791/files#diff-f5e321563d5bc98371a960662f654a9387fd24f9bacb382d5b90d417ad2569c7R39 I solved it by moving the config lookup from import time to runtime. What would you prefer with this?

pradyunsg commented on Lib/ensurepip/__init__.py:
```diff
@@ -105,7 +91,7 @@ def version():
     Returns a string specifying the bundled version of pip.
     """
-    return _get_packages()['pip'].version
+    return _get_pip_version()
```
This can return the constant that is used to locate the pip wheel.

```suggestion
    return _PIP_VERSION
```

pradyunsg commented on Lib/ensurepip/__init__.py:
```diff
@@ -1,78 +1,64 @@
+    )
+
+
+def _get_pip_version():
```
With that change, this function can be removed.

pradyunsg replied:
Ah, we want this to return the `WHEEL_PKG_DIR` version if used.

webknjaz replied:
Yep, I was making sure that the behavior is kept the same..

pradyunsg reviewed (changes requested):
One last change, but LGTM otherwise!

pradyunsg committed:
Merge branch 'main' into maintenance/ensurepip-single-wheel
